### DISCUSSION TITLE
Can't Use With EF - 'No Entity Framework provider found ...'

### DISCUSSION THREAD

**Discussion by BRebey** [mysql-net/MySqlConnector#1353] [Q&A] [answered]
I'm trying to replace MySql.Data.MySqlClient with MySqlConnector in .Net Framework and Web Apps, using EF 6.4.4.  The apps - Console and Web apps alike - target .Net Framework 4.6.1.

Per https://mysqlconnector.net/overview/dbproviderfactories/ , I've made the appropriate replacement, and modified my app to call `DbProviderFactories.GetFactory("MySqlConnector")`.

When I run the app, I get the error 

> InvalidOperationException: No Entity Framework provider found for the ADO.NET provider with invariant name 'MySqlConnector'. Make sure the provider is registered in the 'entityFramework' section of the application config file.  See http://go.microsoft.com/fwlink/?LinkId=260882

If  I add a <provider> section as shown below, I get the error:

> InvalidOperationException: The 'Instance' member of the Entity Framework provider type 'MySqlConnector.MySqlConnectorFactory, MySqlConnector, Version=2.0.0.0, Culture=neutral, PublicKeyToken=d33d3e53aa5f8c92' did not return an object that inherits from 'System.Data.Entity.Core.Common.DbProviderServices'. Entity Framework providers must inherit from this class and the 'Instance' member must return the singleton instance of the provider. This may be because the provider does not support Entity Framework 6 or later; see http://go.microsoft.com/fwlink/?LinkId=260882 for more information.

This suggests to me that the class `MySqlConnector.MySqlConnectorFactory` is wrong.

When using MySql.Data.MySqlClient instead of MySqlConnector, the correct provider appears to be:

```
<provider invariantName="MySql.Data.MySqlClient" type="MySql.Data.MySqlClient.MySqlProviderServices, MySql.Data.EntityFramework, Version=8.0.26.0, Culture=neutral, PublicKeyToken=c5687fc88969c44d"/>
```
(Correct type is 'MySql.Data.MySqlClient.**_MySqlProviderServices_**', which is different  than the type in the associated DbProdiverFactories type of just MySql.Data.MySqlClient)

I don't see anything in the class hierarchy of MySqlConnector that jumps out at me as the correct type to use in the <provider/> tag.

Anyone know what I'm doing wrong here, and how to fix it so that EF6 is happy?

Here is the relevant section of app.config in my .Net Framework app 

(Note that I'm using the same type - MySqlConnector.MySqlConnectorFactory - for both the <provider /> and DbProviderFactories <add /> tags because I don't know what else to use for the <provider /> tag):
```
	<entityFramework>
		<providers>
			<provider invariantName="MySqlConnector" type="MySqlConnector.MySqlConnectorFactory, MySqlConnector, Culture=neutral, PublicKeyToken=d33d3e53aa5f8c92"/>
		</providers>
	</entityFramework>

	<system.data>
		<DbProviderFactories>
			<add name="MySqlConnector"
				invariant="MySqlConnector"
				description="MySQL Connector for .NET"
				type="MySqlConnector.MySqlConnectorFactory, MySqlConnector, Culture=neutral, PublicKeyToken=d33d3e53aa5f8c92" />
		</DbProviderFactories>		
	</system.data>

```

**Comment by bgrainger** [comment, marked as answer]
> This may be because the provider does not support Entity Framework 6 or later

The error message identifies the problem: There is no EF6 support in MySqlConnector. If you're using EF6, the only option is `MySql.Data.EntityFramework` with `MySql.Data`. (As far as I know, there is no way to use the Oracle EF implementation with the MySqlConnector ADO.NET implementation.)

If you can migrate to EF Core, then you can use [Pomelo.EntityFrameworkCore.MySql](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql) which supports EF Core 7. (Note that EF7 requires .NET 6, but EF Core 3.1.x can run on .NET Framework 4.6.1; however, it may be unsupported.)
